### DISCUSSION TITLE
fastapi-HTTPException works correctly with graphene?

### DISCUSSION THREAD

**Discussion by danielahdez** [fastapi/fastapi#7710] [Questions] [answered]
Hi everyone. I am currently working on a graphql-Python service layer using fastapi web framework. Right now I'm developing the security part and need to handle specific http exceptions. So I have been reading the official documentation of fastapi to handle this type of exceptions using REST and when I include them in my GraphQL project I am not getting the expected results.

This is one of the methods where I need it (still in process lol): 
![E9719B66-2949-49B5-A09B-8220110F40CF](https://user-images.githubusercontent.com/58786990/78820547-fed29600-79a5-11ea-8499-3631000672fe.jpeg)

And this is my main (just in case):
![F2E09A3D-3C4C-4584-A326-41FD89D069B8](https://user-images.githubusercontent.com/58786990/78820571-0abe5800-79a6-11ea-8d7f-e4b480aba2b3.jpeg)

This is the response from the uvicorn server which returns a status 400 even if I am asking for a 403... The server never handles the exception:
![4A53A6FB-CD90-4A31-8D2B-C0397DD1AD47](https://user-images.githubusercontent.com/58786990/78820601-1ad63780-79a6-11ea-8c82-88608c3db2b9.jpeg)

and this is the view of the response from the client side (it doesn't even show the error message)
![919F9A57-9DBC-4F5B-92EB-742730D202C6](https://user-images.githubusercontent.com/58786990/78820632-2295dc00-79a6-11ea-92ee-895993d5775c.jpeg)

I just want to know if this kind of exceptions works perfectly for graphene or do I need something more or different ..... or I misunderstood something and am not doing it correctly?

**Comment by tiangolo** [comment, marked as answer]
Hi @danielahdez ! 

So, the `HTTPException` works when you use the normal API parts of FastAPI, using things like `@app.get("/")` and `@app.post("/users/")`, etc.

When you use GraphQL, it handles everything internally, how to declare data shapes/schemas, how to read them and process them, how to retrieve data from a database, return errors, etc.

When using GraphQL, it all depends mainly on the system you are using for GraphQL, like Graphene or Ariadne, not so much on FastAPI.

You could still have other normal endpoints at other paths in your server with normal JSON, and handle them with normal FastAPI code, as you can see in the docs and the tutorial. And have that in addition to all your GraphQL app.

But when using GraphQL, the library that you choose ends up handling more or less everything.

For example, in Graphene you would handle errors like this: https://www.howtographql.com/graphql-python/6-error-handling/

Or if you were using Ariadne, you would integrate it like this: https://ariadnegraphql.org/docs/starlette-integration.html

**Comment by danielahdez** [comment]
Hi @tiangolo! 

I've already been using Graphene error handling for other types of exceptions for a while. This was my first choice, but after a few days of researching and trying various options I couldn't find a way to change the status of an HTTP response as I wanted (401/403 status) ...

So that's why I decided to try using Fastapi and Uvicorn `HTTPException`... But you're right, GraphQL usually handles everything internally, his internal structure does not allow this type of third-party management. He keeps interrupting those third-party exceptions and running its own exceptions returning a `400 Bad Request` no matter what. 

I've been reading and don't think there is a clear way to change HTTP states in Graphene-GraphQL for now (at least I haven't found any way yet).

Thank you very much for your help! 🍫 and by the way I love working with FastApi, nice job!🍫

Sorry my English. Closing this issue.

**Comment by tiangolo** [comment]
Thanks for coming back to report and close the issue! :+1: 

Sorry I can't help more, I'm not an expert with Graphene :sweat_smile: 

Another option, although it's a big change might be to try Ariadne...

I understand in many GraphQL libraries they return the same error code for everything and another internal more specific error code in the response, that might be what they are expecting :shrug: 

I'm glad you're liking FastAPI! :smile: :coffee: 

> Sorry my English. Closing this issue.

No worries, sorry for mine.

...Podríamos haber estado hablando español todo el tiempo, pero gracias por mantenerlo en inglés, así otros se pueden beneficiar de la conversación luego :smile: 

Good luck!

**Comment by mcauto** [comment]
@danielahdez Did you solve it?

**Comment by danielahdez** [comment]
@mcauto not really, I ended up implementing regular Exceptions.

As I explained in my previous comment, in the backend when you try to execute HTTP Exceptions, Graphene interrupts these Exceptions and returns its own, thats means for example that you may want to throw a 401 or 403 Exception when the user tries to make a request to the server and is not Authenticated/Authorized but when Graphene processes the request it will interrupt this and show always a status of 400 in response. 

I tried multiple HTTP libraries and the most it could get with some of them was the Exception message, the state never changed.

It’s not an error, it is the way that Graphene works and processes requests.
